File: LinkedList/Easy/LinkedListPalindrome/palindrome.cpp
#include <iostream>
using namespace std;

/* Node class for the linked list */
class VSDnode {
public:
    int data;
    VSDnode* next;

    /* Constructor to initialize node with given value */
    VSDnode(int val) {
        data = val;
        next = NULL;
    }
};

/* Function to add a new node at the head of the linked list */
void add_head(VSDnode* &head, int val) {
    VSDnode* n = new VSDnode(val);
    n->next = head;
    head = n;
}

/* Function to add a new node at the tail of the linked list */
void add_tail(VSDnode* &head, int val) {
    VSDnode* n = new VSDnode(val);
    if (head == NULL) {
        head = n;
        return;
    }
    VSDnode* temp = head;
    while (temp->next != NULL) {
        temp = temp->next;
    }
    temp->next = n;
}

/* Function to display the linked list */
void display(VSDnode* head) {
    VSDnode* temp = head;
    while (temp != NULL) {
        cout << temp->data << "->";
        temp = temp->next;
    }
    cout << "NULL" << endl;
}

/* Function to reverse the linked list */
VSDnode* reverse(VSDnode* head) {
    VSDnode* prev = NULL;
    VSDnode* curr = head;

    while (curr) {
        VSDnode* next = curr->next;
        curr->next = prev;
        prev = curr;
        curr = next;
    }
    return prev;
}

/* Function to check if the linked list is a palindrome */
void is_palindrome(VSDnode* head) {
    VSDnode* new_head = reverse(head);
    VSDnode* temp1 = head;
    VSDnode* temp2 = new_head;

    while (temp1) {
        if (temp1->data != temp2->data) {
            cout << "The linked list is not a palindrome" << endl;
            return;
        }
        temp1 = temp1->next;
        temp2 = temp2->next;
    }
    cout << "The linked list is a palindrome" << endl;
}

int main() {
    cout << "Enter the number of elements in the linked list: ";
    int n;
    cin >> n;

    VSDnode* head = NULL;

    cout << "Enter the elements of the linked list:" << endl;
    for (int i = 0; i < n; ++i) {
        int val;
        cin >> val;
        add_tail(head, val);
    }

    cout << "Linked List:" << endl;
    display(head);

    /* Check if the linked list is a palindrome */
    is_palindrome(head);

    return 0;
}

File: LinkedList/Easy/LinkedListPalindrome/palindrome.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

def create_list(llist):
    head = None
    temp = None
    for i in llist:
        if not temp:
            temp = Node(i)
            head = temp
        else:
            temp.next = Node(i)
            temp = temp.next
    return head

def check_palindrome(head):
    temp = head
    stack = []

    # Push elements of the first half onto the stack
    while head:
        stack.append(head.data)
        head = head.next

    # Compare the second half with the elements popped from the stack
    while temp:
        if temp.data != stack.pop():
            return False
        temp = temp.next

    return True

if __name__ == "__main__":
    # Take user input for the linked list
    user_input = input("Enter elements of the linked list separated by spaces: ")
    llist = list(map(int, user_input.split()))

    # Create the linked list
    head = create_list(llist)

    # Check if the linked list is a palindrome
    result = check_palindrome(head)

    # Print the result
    if result:
        print("The linked list is a palindrome.")
    else:
        print("The linked list is not a palindrome.")
